src/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
```

Comment:
this is quite a useful chunk of funcs, at the same time it adds quite a lot of complexity

is the plan to offload these to a separate library when we have estabilished this is the path?

reason is this is quite complex and it's basically adding a level of complexity and abstraction on top of what we already have in the `ops` lib

it would require a thorough documentation on how to use and hook up something into the codebase, without that it is quite complicated to maintain in each charm codebase and easy to ahck when things don't work as expected

i'm more than happy to keep it, but we need tons of docs to explain how to use it and how to fix this oin case stuff needs to be enhanced or changed, basically it requires a library on its own and a set of team maintainers imho

Reply:
I am not sure if there is a plan to put those into a lib at this point. This needs to be decided by the team.

The original initiative comes from what I did in glauth-k8s. I saw lots of duplicated code sitting in different event handlers. So I thought it would be nice to reduce the duplications. The validations and conditions are pretty common in every charm based on what I observed so far.

I can remove those decorators if they are too complex.